---
title: "Analysis of outlier samples"
author: "Jared J. Homola"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
Multidimensional scaling plots for all sequenced mallard bursa and ileum samples.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
## Load packages and data
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(RColorBrewer)
library(gridExtra)
library(kableExtra)
library(rrcov)
library(ggrepel)
library(tidyverse)

## Load data
cnt.bursa <- read.table("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/all.genes.results.csv", header = TRUE) %>% as_tibble(rownames = NA)
cnt.ileum <- read.csv("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/adjusted_read_counts.csv", header = TRUE) %>%
  as_tibble(rownames = NA)
covars <- read_delim("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/MALL70_SSgroup_RAW_7.7.20.csv", delim = ",")
annot.bursa <- read.delim("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/annotationGenes102320.txt", header = TRUE, sep = "\t")
annot.ileum <- cnt.ileum$gene
```


### Clean up data
```{r, warning = FALSE, message = FALSE}
### Bursa
newNames.bursa <- names(cnt.bursa) %>%
  as_tibble() %>%
  separate(value, into = c("pt1", "pt2", "pt3")) %>%
  select(pt2, pt3) %>%
  mutate(pt2 = as.numeric(pt2),
         pt2 = formatC(pt2, width = 2, format = "d", flag = "0")) %>%
  mutate(newName = paste0("bursa_", pt2))

colnames(cnt.bursa) <- newNames.bursa$newName

rownames.bursa <- annot.bursa %>%
  mutate(rowname = ifelse(hgnc_symbol == "", ensembl_gene_id, hgnc_symbol)) %>%
  add_count(rowname) %>%
  mutate(rowname = ifelse(n < 2, rowname, paste0(rowname, "_", row_number())))

cnt.bursa.tib <- cnt.bursa[,order(colnames(cnt.bursa))] %>%
  mutate(gene = rownames.bursa$rowname) %>%
  column_to_rownames("gene") %>%
  as_tibble(rownames = NA)


### Ileum
cnt.ileum.tib <- cnt.ileum %>%
  as_tibble() %>%
  select(-chr, -gene)

newNames.ileum <- names(cnt.ileum.tib) %>%
  as_tibble() %>%
  mutate(value = gsub("Iluem_", "", .$value)) %>% 
  separate(value, into = c(NA, "pt1", "pt2")) %>%
  mutate(pt1 = na_if(pt1, "MALL"),
         pt2 = na_if(pt2, "ileum")) %>%
  mutate(bird = ifelse(is.na(pt1), pt2, pt1)) %>%
  mutate(bird = as.numeric(bird),
    bird = formatC(bird, width = 2, format = "d", flag = "0")) %>%
  mutate(newName = paste0("ileum_", bird))

colnames(cnt.ileum.tib) <- newNames.ileum$newName

cnt.ileum.tib <- cnt.ileum.tib[,order(colnames(cnt.ileum.tib))] %>%
  mutate(gene = cnt.ileum$gene) %>%
  column_to_rownames("gene") %>%
  as_tibble(rownames = NA)

### Covars
covars <- covars %>%
  mutate(bird = as.numeric(bird)) %>%
  mutate(bird = formatC(bird, width = 2, format = "d", flag = "0")) %>%
  arrange(bird) %>%
  mutate(bird = as.factor(bird)) %>%
  mutate(group = str_remove(group, "-"))

covars.ileum <- covars %>%
  filter(bird %in% newNames.ileum$bird)
```

### Remove birds as needed
```{r, warning = FALSE, message = FALSE}
cnt.bursa.tib <- cnt.bursa.tib %>% select(-bursa_01,
                                          -bursa_72)
cnt.ileum.tib <- cnt.ileum.tib %>% select(-ileum_01,
                                          -ileum_72)

covars <- covars %>% filter(bird != "01",
                            bird != "72")

covars.ileum <- covars.ileum %>% filter(bird != "01",
                                        bird != "72")
```

### Prep data for analysis
Here we standardize expression levels across samples of varying depth by using the counts per million (CPM) and trimmed mean of M-values (TMM) methods. We also remove any genes that are expressed at few than 0.5 CPM in at least 25% of individuals.
```{r, warning = FALSE, message = FALSE}
### Filter out lowly expressed transcripts
#Convert to DGEList object
dge.bursa <- DGEList(counts=cnt.bursa.tib)

#CPM and log-CPM
cpm.bursa <- cpm(dge.bursa)
lcpm.bursa <- cpm(dge.bursa, log = TRUE)

#### Retain only genes expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge.bursa$counts == 0) == length(cnt.bursa.tib[1,]))
keep.exprs.bursa <- rowSums(cpm.bursa > 0.5) >= length(cnt.bursa.tib[1,])/4
sum(keep.exprs.bursa)

dge.bursa <- dge.bursa[keep.exprs.bursa,, keep.lib.sizes=FALSE]
table(rowSums(dge.bursa$counts==0) == length(cnt.bursa.tib[1,]))
dim(dge.bursa)

#TMM normalization
dge.bursa <- calcNormFactors(dge.bursa, method = "TMM")


#Convert to DGEList object
dge.ileum <- DGEList(counts=cnt.ileum.tib)

#CPM and log-CPM
cpm.ileum <- cpm(dge.ileum)
lcpm.ileum <- cpm(dge.ileum, log = TRUE)

#### Retain only genes expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge.ileum$counts == 0) == length(cnt.ileum.tib[1,]))
keep.exprs.ileum <- rowSums(cpm.ileum > 0.5) >= length(cnt.ileum.tib[1,])/4
sum(keep.exprs.ileum)

dge.ileum <- dge.ileum[keep.exprs.ileum,, keep.lib.sizes=FALSE]
table(rowSums(dge.ileum$counts==0) == length(cnt.ileum.tib[1,]))
dim(dge.ileum)

#TMM normalization
dge.ileum <- calcNormFactors(dge.ileum, method = "TMM")
```

### Plot MDS
#### Tissue groupings
##### Purpose: Is there evidence of tissue misassignment?
```{r, warning = FALSE, message = FALSE}
lcpm.bursa2 <- lcpm.bursa %>% as.data.frame() %>% rownames_to_column("gene")
lcpm.ileum2 <- lcpm.ileum %>% as.data.frame() %>% rownames_to_column("gene")

lcpm.all <- right_join(as.data.frame(lcpm.bursa2), as.data.frame(lcpm.ileum2))
tissueCode <- colnames(lcpm.all) %>% as_tibble() %>% separate(value, into = c("tissue", NA)) %>% slice(-1)

col.group = as.numeric(as.factor(tissueCode$tissue))
plotMDS(lcpm.all[-1], labels = paste0(colnames(lcpm.all[-1]), "_", tissueCode$tissue), col = col.group, cex = 0.75)
title(main="Tissues")
```

#### Bursa groupings
##### Purpose: Are there any outliers that should be removed?
```{r, warning = FALSE, message = FALSE}
col.group = as.numeric(as.factor(covars$group))
plotMDS(lcpm.bursa, labels = paste0(colnames(lcpm.bursa), "_", covars$group), col = col.group, cex = 0.75)
title(main="Treatment group - Bursa")

col.group = as.numeric(as.factor(covars$sex))
plotMDS(lcpm.bursa, labels = paste0(colnames(lcpm.bursa), "_", covars$sex), col = col.group, cex = 0.75)
title(main="Sex - Bursa")

col.group = as.numeric(as.factor(covars$Bursa_pool))
plotMDS(lcpm.bursa, labels = paste0(colnames(lcpm.bursa), "_", covars$Bursa_pool), col = col.group, cex = 0.75)
title(main="Pool - Bursa")
```

#### Ileum groupings
##### Purpose: Are there any outliers that should be removed?
```{r, warning = FALSE, message = FALSE}
col.group = as.numeric(as.factor(covars.ileum$group))
plotMDS(lcpm.ileum, labels = paste0(colnames(lcpm.ileum), "_", covars.ileum$group), col = col.group, cex = 0.75)
title(main="Treatment group - Ileum")

col.group = as.numeric(as.factor(covars.ileum$sex))
plotMDS(lcpm.ileum, labels = paste0(colnames(lcpm.ileum), "_", covars.ileum$sex), col = col.group, cex = 0.75)
title(main="Sex - Ileum")

col.group = as.numeric(as.factor(covars.ileum$Bursa_pool))
plotMDS(lcpm.ileum, labels = paste0(colnames(lcpm.ileum), "_", covars.ileum$Ileum_pool), col = col.group, cex = 0.75)
title(main="Pool - Ileum")
```



#### Robust PCA outlier maps
The original description of this method of multivariate outlier detection is given by Hubert and Engelsen 2004 (Bioinformatics 20:1728-1736). It was implemented for RNA-seq data by Chen et al. 2020 (BMC Bioinformatics 21:269). A description of the plot interpretation is below. This describes each of the quadrants of the plots (taken from Hubert and Engelsen 2004):  
  
* Lower left quadrant: Regular  observations have  a  small  orthogonal  and  a small score distance.  
* Lower right quadrant: When samples have a large score distance, but a small orthogonal distance, we call them good leverage points.  These observations lie close to the space spanned by the principal components but far from the regular data.  This implies that they are different from the majority,  but there is only a little loss of information when we replace them by their fitted values in the PCA-subspace.  
* Upper right quadrant: Orthogonal outliers have a large orthogonal distance, but a small score distance. They  can  not  be  distinguished  from  the  regular observations  once  they  are  projected  onto  the  PCA  sub-space, but they lie far from this subspace. Consequently,it would be dangerous to replace that sample with its projected value, as its outlyingness would not be visible anymore.  
* Upper left quadrant: Bad leverage points have a large orthogonal distance and a large score distance. They  lie  far  outside  the  space  spanned  by  the  principal components, and after projection far from the regular data points. Their degree of outlyingness is high in both directions, and typically they have a large influence on CPCA,as the eigenvectors will be tilted towards them.
```{r, warning = FALSE, message = FALSE}
### How many axes to retain?
pca.bursa <- prcomp(lcpm.bursa, scale. = TRUE)
varExplained.bursa <- data.frame(PC= paste0("PC",1:68),
                               var_explained=(pca.bursa$sdev)^2/sum((pca.bursa$sdev)^2)) %>%
  slice(1:5)

pca.ileum <- prcomp(lcpm.ileum, scale. = TRUE)
varExplained.ileum <- data.frame(PC= paste0("PC",1:42),
                               var_explained=(pca.ileum$sdev)^2/sum((pca.ileum$sdev)^2)) %>%
  slice(1:5)


varExplained.bursa %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)

varExplained.ileum %>%
  kable() %>%
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)
```



```{r, warning = FALSE, message = FALSE}
### How many axes to retain?
pca.ileum <- PcaGrid(t(lcpm.ileum), k = 2)
sd.ileum <- pca.ileum@sd %>% as_tibble() %>% rename(scoreDistance = value)
od.ileum <- pca.ileum@od %>% as_tibble() %>% rename(orthoDistance = value)
plotDat.ileum <- bind_cols(sd.ileum, od.ileum) %>%
  mutate(sample = rownames(pca.ileum@scores))

ggplot(plotDat.ileum, aes(x = scoreDistance, y = orthoDistance, label = sample)) +
  geom_point() +
  xlab("Score distance") +
  ylab("Orthogonal distance") +
  geom_text_repel(data = subset(plotDat.ileum, orthoDistance > pca.ileum@cutoff.od)) +
  geom_hline(yintercept = pca.ileum@cutoff.od) +
  geom_vline(xintercept = pca.ileum@cutoff.sd) +
  ggtitle("Outlier map - Ileum") +
  theme_bw(base_size = 14)


pca.bursa <- PcaGrid(t(lcpm.bursa), k = 2)
sd.bursa <- pca.bursa@sd %>% as_tibble() %>% rename(scoreDistance = value)
od.bursa <- pca.bursa@od %>% as_tibble() %>% rename(orthoDistance = value)
plotDat.bursa <- bind_cols(sd.bursa, od.bursa) %>%
  mutate(sample = rownames(pca.bursa@scores))

ggplot(plotDat.bursa, aes(x = scoreDistance, y = orthoDistance, label = sample)) +
  geom_point() +
  xlab("Score distance") +
  ylab("Orthogonal distance") +
  geom_text_repel(data = subset(plotDat.bursa, orthoDistance > pca.bursa@cutoff.od)) +
  geom_hline(yintercept = pca.bursa@cutoff.od) +
  geom_vline(xintercept = pca.bursa@cutoff.sd) +
  ggtitle("Outlier map - Bursa") +
  theme_bw(base_size = 14)
```
