---
title: "Shed Level I2 (LvMvH) - Ileum - Gene"
author: "Amanda Dolinski & Jared J. Homola"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  pdf_document: default
  html_document: default
  word_document: default
---

\pagenumbering{gobble}

This is an analysis of differential expression at the gene level between low, moderate, and high LPAIV shedding mallards based on cloacal swab virus titers measured on the day of sacrifice for ileum samples on 2 days post infection (DPI).

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r message=FALSE, warning=FALSE}
## Load packages and data
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(ggrepel)
library(RColorBrewer)
library(gridExtra)
library(kableExtra)
library(tidyverse)

## Load data
cnt1 <- read_delim("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/mallardLPAIV/extData/ileumCounts.txt", delim = "\t") %>% distinct(gene, .keep_all = TRUE)
covars <- read_delim("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/mallardLPAIV/extData/MALL70_SSgroup_RAW_12.3.20.csv", delim = ",")
annot <- read_delim("G:/Shared drives/RNA Seq Supershedder Project/MALL DE manuscript/mallardLPAIV/extData/annotationGenes103120.txt", delim = "\t")
```

```{r, warning = FALSE, message = FALSE}
cnt <- cnt1 %>%
  as_tibble() %>%
  select(-gene)

newNames.ileum <- names(cnt) %>%
  as_tibble() %>%
  mutate(value = gsub("Iluem_", "", .$value)) %>% 
  separate(value, into = c(NA, "pt1", "pt2")) %>%
  mutate(pt1 = na_if(pt1, "MALL"),
         pt2 = na_if(pt2, "ileum")) %>%
  mutate(bird = ifelse(is.na(pt1), pt2, pt1)) %>%
  mutate(bird = as.numeric(bird),
    bird = formatC(bird, width = 2, format = "d", flag = "0")) %>%
  mutate(newName = paste0("ileum_", bird))

colnames(cnt) <- newNames.ileum$newName

cnt <- cnt[,order(colnames(cnt))]  %>%
  as_tibble(rownames = NA)

### Covars
covars3 <- covars %>%
  mutate(bird = as.numeric(bird)) %>%
  mutate(bird = formatC(bird, width = 2, format = "d", flag = "0")) %>%
  arrange(bird) %>%
  mutate(bird = as.factor(bird)) %>%
  mutate(group = str_remove(group, "-"))

covars.ileum <- covars3 %>%
  filter(bird %in% newNames.ileum$bird)

cnt <- cnt %>% 
  select(-ileum_01, -ileum_72) %>%
  mutate(gene = cnt1$gene) %>%
  column_to_rownames("gene")

covars.ileum <- covars.ileum %>% filter(bird != "01",
                                        bird != "72")
```

```{r, warning = FALSE, message = FALSE, results = 'hide'}
#Convert to DGEList object
dge <- DGEList(counts=cnt)
dge$genes <- annot[match(rownames(dge$counts), annot$ensembl_gene_id),]

#CPM and log-CPM
cpm <- cpm(dge)
lcpm <- cpm(dge, log = TRUE)

#### Retain only genes expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge$counts==0) == length(cnt[1,]))
keep.exprs <- rowSums(cpm>0.5) >= length(cnt[1,])/4
sum(keep.exprs)

dge <- dge[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge$counts==0) == length(cnt[1,]))
dim(dge)
```

```{r, warning = FALSE, message = FALSE}
#Assign groups in DGElist
group <- recode(covars.ileum$group, C1 = "Ctl", C29 = "Ctl") %>% 
  as.factor()
dge$samples$group <- group

### Subset to I2
dge.I2 <- dge[,grep('\\bI2\\b', dge$samples$group)]
dge.I2 <- calcNormFactors(dge.I2, method = "TMM")

covars.I2 <- covars3 %>% filter(group == "I2")
```


# Differential expression analysis
```{r}
## Establish design matrix: I2
design.I2 = model.matrix(
  ~ 0 +
    factor(covars.I2$SSgroup.virus.sac) +
    covars.I2$sex +
    covars.I2$Ileum_pool
)

colnames(design.I2) <- c(
  "HIGH",
  "LOW",
  "MODERATE",
  "sexM",
  "poolMSU6",
  "poolUMN"
)

contr.matrix.I2 = makeContrasts(
  LvM = LOW - MODERATE,
  MvH = MODERATE - HIGH,
  LvH = LOW - HIGH,
  levels = colnames(design.I2))

## Mean-variance trend plots
v.I2 <- voomWithQualityWeights(dge.I2, design.I2, plot = FALSE)

#Fitting models
vfit.I2 <- lmFit(v.I2, design.I2) 
vfit.I2 <- contrasts.fit(vfit.I2, contrasts = contr.matrix.I2)
tfit.I2 <- treat(vfit.I2, lfc = 0.5)

## Identify DE genes
dt.I2 <- decideTests(tfit.I2, p.value = 0.1, adjust.method = "fdr")
dt.tib <- as_tibble(dt.I2, rownames = NA) %>% 
  rownames_to_column("gene") %>% 
  mutate_at(vars(starts_with("L")), as.numeric) %>% 
  mutate_at(vars(starts_with("M")), as.numeric) %>% 
  rename(LvM.I2 = LvM,
         MvH.I2 = MvH,
         LvH.I2 = LvH) %>% 
  filter_at(vars(LvM.I2:LvH.I2), any_vars(. != 0))

allResults <- rbind(
  as.data.frame(summary(dt.I2))) %>% 
  filter(Var1 != "NotSig") %>% 
  rename("Direction" = Var1,
         "Comparison" = Var2,
         "N" = Freq)

kbl(allResults, booktabs = T, caption = "Count of DE genes. For a gene to be considered differentially expressed, we require a p-value of 0.1 with a false discovery rate correction and a log fold change difference of 0.5.") %>% 
  kable_styling(position ="center", latex_options = c("hold_position"))
```

\newpage 

### Volcano plot
Volcano plot reporting -log10(p-values) as a function of log2(fold change) between the samples (logFC, x axis). Transcripts/genes that are identified as significantly differentially expressed following a false discovery rate correction (q = 0.10) are shown in red
```{r, warning = FALSE, message = FALSE, fig.height = 5, fig.width = 9, fig.align = "center"}
tmp1 <- topTreat(tfit.I2, coef = 1, n = Inf)
results1 <- mutate(tmp1, sig=ifelse(tmp1$adj.P.Val<0.1, "Sig", "Not Sig"))
p1 <- ggplot(results1, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. Moderate shedding, Day 2") +
  ylab("-Log10(p-value)") +
  xlab("Log(fold change)") +
  theme_bw() +
  theme(legend.position = "none")

tmp2 <- topTreat(tfit.I2, coef = 2, n = Inf)
results2 <- mutate(tmp2, sig=ifelse(tmp2$adj.P.Val<0.1, "Sig", "Not Sig"))
p2 <- ggplot(results2, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Moderate vs. High shedding, Day 2") +
  ylab("-Log10(p-value)") +
  xlab("Log(fold change)") +
  theme_bw() +
  theme(legend.position = "none")

tmp3 <- topTreat(tfit.I2, coef = 3, n = Inf)
results3 <- mutate(tmp3, sig=ifelse(tmp3$adj.P.Val<0.1, "Sig", "Not Sig"))
p3 <- ggplot(results3, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Low vs. High shedding, Day 2") +
  ylab("-Log10(p-value)") +
  xlab("Log(fold change)") +
  theme_bw() +
  theme(legend.position = "none")

grid.arrange(p1, p2, p3, nrow = 1)
```

### Heatmap
Hierarchical clustering of transcripts/genes and samples. Shown is a heat map illustrating the relative expression levels of each transcript (rows) in each sample (column). Rows are hierarchically clustered by expression. Log2-transformed expression values are z-transformed.
```{r warning = FALSE, message = FALSE, fig.height = 40, fig.width = 9, fig.align = "center"}
# Get the gene names for DE genes
lcpm2 <- as_tibble(dge.I2$counts, rownames = NA) %>% 
  cpm(., log = TRUE)

newNames <- colnames(lcpm2) %>% 
  as_tibble() %>% 
  separate(value, into = c(NA, "bird")) %>% 
  left_join(covars.I2, by = "bird") %>% 
  unite("newNames", SSgroup.virus.sac, group, bird)

colOrder <- bind_cols(covars.I2) %>% 
  mutate(SSgroup.virus.sac = fct_relevel(SSgroup.virus.sac, "LOW", "MODERATE", "HIGH")) %>%
  arrange(SSgroup.virus.sac, group, bird) %>% 
  unite("newNames", SSgroup.virus.sac, group, bird)

colnames(lcpm2) <- newNames$newNames
lcpm2 <- lcpm2[, colOrder$newNames]

heatmap.annot <- annot %>%
  filter(ensembl_gene_id %in% dt.tib$gene) %>% 
  mutate(hgnc_symbol = replace_na(hgnc_symbol, ".")) %>% 
  unite(annot.hm, c(ensembl_gene_id, hgnc_symbol), sep = "-", remove = FALSE) %>% 
  select(annot.hm, ensembl_gene_id) %>% 
  distinct(ensembl_gene_id, .keep_all = TRUE)

deGeneCounts <- lcpm2 %>% 
  as_tibble() %>% 
  mutate(ensembl_gene_id = rownames(dge.I2$counts)) %>% 
  filter(ensembl_gene_id %in% dt.tib$gene) %>% 
  left_join(heatmap.annot) %>% 
  column_to_rownames("annot.hm") %>%
  select(-ensembl_gene_id) %>% 
  as.matrix()

## Set up palette
mypalette <- brewer.pal(11,"RdYlBu")
morecols <- colorRampPalette(mypalette)

hc <- hclust(as.dist(1-cor(t(deGeneCounts))))

# Plot the heatmap
heatmap.2(deGeneCounts,
          Colv = FALSE,
          Rowv = as.dendrogram(hc),
          col = rev(morecols(50)),
          trace = "none",
          colsep = c(4, 8),
          dendrogram = "row",
          density.info = "none",
          key = FALSE,
          margins = c(12, 12),
          scale ="row",
          lwid=c(2.5,4), 
          lhei=c(0.1,4))
```

```{r message=FALSE, warning=FALSE}
tmp1a <- tmp1 %>% select(ensembl_gene_id, adj.P.Val, gene_biotype, logFC, hgnc_symbol, description) %>% mutate(comp = "LvM.I1")
tmp2a <- tmp2 %>% select(ensembl_gene_id, adj.P.Val, gene_biotype, logFC, hgnc_symbol, description) %>% mutate(comp = "MvH.I1")
tmp3a <- tmp3 %>% select(ensembl_gene_id, adj.P.Val, gene_biotype, logFC, hgnc_symbol, description) %>% mutate(comp = "LvH.I1")


bind_rows(tmp1a, tmp2a, tmp3a) %>% 
  select(-description) %>% 
  filter(ensembl_gene_id %in% dt.tib$gene) %>% 
  mutate(logFoldCount = ifelse(adj.P.Val < 0.10, round(logFC, 2), "ns")) %>% 
  select(-adj.P.Val, -logFC, -gene_biotype) %>% 
  pivot_wider(names_from = comp, values_from = logFoldCount) %>% 
  kbl(booktabs = T, 
      longtable =T,
      caption ="Annotations for differentially expressed genes. ns denotes non-significant genes for each comparison and numerical values are the log(fold change) difference") %>%
  kable_styling(latex_options = c("repeat_header"))
```

```{r message=FALSE, warning=FALSE}
bind_rows(tmp1a, tmp2a, tmp3a) %>% 
  select(-comp, -adj.P.Val, -logFC) %>% 
  filter(ensembl_gene_id %in% dt.tib$gene) %>% 
  distinct(.) %>%
  select(-ensembl_gene_id) %>% 
  kbl(booktabs = TRUE, 
      longtable =T,
      caption = "Gene functions") %>%
  kable_styling(latex_options = c("repeat_header")) %>% 
  landscape() %>% 
  column_spec(4, width ="30em")
  #kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```

```{r message=FALSE, warning=FALSE}
de.entrez <- annot %>% filter(ensembl_gene_id %in% dt.tib$gene, 
                              entrezgene_id > 0)
entrez.universe <- annot %>% na.omit(entrezgene_id)

K <- kegga(de.entrez$entrezgene_id, universe = entrez.universe$entrezgene_id, species.KEGG = "apla") %>% 
  as_tibble(rownames = NA) %>% 
  rownames_to_column("PathwayID") %>% 
 # filter(P.DE < 0.05) %>% 
  arrange(P.DE)

keggList <- getGeneKEGGLinks(species.KEGG = "apla") %>% 
  as_tibble() %>% 
  rename("entrezgene_id" = GeneID)

keggGenes = data.frame()
for (path in nrow(K)){
    procPath <- filter(keggList, entrezgene_id %in% de.entrez$entrezgene_id)
    df <- data.frame(procPath)
    keggGenes <- rbind(keggGenes, df)
}

left_join(K, keggGenes) %>%
  as_tibble() %>%
  mutate(entrezgene_id = as.numeric(as.character(entrezgene_id))) %>% 
  left_join(annot) %>%  
  select(Pathway, PathwayID, N, DE, P.DE, ensembl_gene_id, hgnc_symbol, entrezgene_id) %>%
  filter(DE > 0) %>% 
  kbl(booktabs = T, 
      longtable = T,
      caption = "KEGG pathway analysis: pathways with p-values <0.05 were determined as over-represented (enriched) pathways of differentially expressed genes/transcripts") %>%
  kable_styling(latex_options =c("repeat_header")) %>% 
  landscape() %>% 
  column_spec(1, width ="15em") %>% 
  column_spec(6, width ="10em")
  #kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
```
