---
title: "Treatment Groups - Bursa - Gene"
author: "Jared J. Homola"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---
This is an analysis of differential expression at the gene level between treatment groups for control and infected mallard bursa samples.

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r message=FALSE, warning=FALSE}
## Load packages and data
library(limma)
library(edgeR)
library(Glimma)
library(gplots)
library(ggrepel)
library(RColorBrewer)
library(gridExtra)
library(kableExtra)
library(topGO)
library(tidyverse)

## Load data
cnt <- read.csv("../../degnorm_08272020_162816/adjusted_read_counts.csv", header = TRUE) %>% as_tibble() %>% 
  select(-chr)
covars <- read_delim("../../MALL70_SSgroup_RAW_7.7.20.csv", delim = ",")
annot <- read.delim("../../annotationGenes52820.txt", header = TRUE, sep = "\t")
```


### Clean data
```{r, warning = FALSE, message = FALSE}
colNames <- colnames(cnt)[-1] %>% 
  as_tibble() %>% 
  mutate(name = gsub("Aligned.out.sorted", "", .$value)) %>% 
  mutate(name = gsub("ileum", "", .$name)) %>% 
  mutate(name = gsub("mallard__", "", .$name)) %>% 
  mutate(name = gsub("mallard_", "", .$name)) %>% 
  separate(name, "bird") %>%
  mutate(bird = str_pad(bird, 2, pad = "0")) %>% 
  mutate(colName = paste0("ileum_", bird))

colnames(cnt) <- c("gene", colNames$colName)

cnt <- cnt[,order(colnames(cnt))] %>% 
    column_to_rownames("gene")

covars <- covars %>%
  mutate(bird = as.numeric(bird)) %>%
  mutate(bird = str_pad(bird, 2, pad = "0")) %>% 
  arrange(bird) %>%
  mutate(bird = as.factor(bird)) %>% 
  mutate(group = str_remove(group, "-")) %>% 
  drop_na(RINileum_seq)

### Annotation information isn't aligning right
annot2 <- annot %>% 
  mutate(annotation = coalesce(hgnc_symbol, ensembl_gene_id)) %>% 
  head()
```

### Set up model effects
```{r, warning = FALSE, message = FALSE}
bird <- as.factor(covars$bird)
sex <- as.factor(covars$sex)
age <- as.numeric(covars$age)
weight <- covars$wt_55
group <- recode(covars$group, C1 = "Ctl", C29 = "Ctl") %>% 
  as.factor()
pool <- as.factor(covars$Ileum_pool)
```

### Prep data for analysis
Here we standardize expression levels across samples of varying depth by using the counts per million (CPM) and trimmed mean of M-values (TMM) methods. We also remove any genes that are expressed at few than 0.5 CPM in at least 25% of individuals.
```{r, warning = FALSE, message = FALSE}
#Convert to DGEList object
dge <- DGEList(counts=cnt)
dge$genes <- annot[match(rownames(dge$counts), annot$hgnc_symbol),]

#CPM and log-CPM
cpm <- cpm(dge)
lcpm <- cpm(dge, log = TRUE)

#### Retain only genes expressed at >0.5 CPM in at least 25% of individuals
table(rowSums(dge$counts==0) == length(cnt[1,]))
keep.exprs <- rowSums(cpm>0.5) >= length(cnt[1,])/4
sum(keep.exprs)

dge <- dge[keep.exprs,, keep.lib.sizes=FALSE]
table(rowSums(dge$counts==0) == length(cnt[1,]))
dim(dge)

#TMM normalization
dge <- calcNormFactors(dge, method = "TMM")
```

### Plot MDS
Multidimensional scaling (MDS) plots are an ordination technique that lets us examine sample clustering based on overall gene expression levels.
```{r}
col.group = as.numeric(as.factor(paste0(group)))
plotMDS(lcpm, labels = paste0(bird, "_", group), col = col.group, cex = 0.75)
title(main="Treatment group")
```

# Differential expression analysis
### Establish design & contrast matrices
```{r}
## Establish design matrix
design = model.matrix(~ 0 +
                        group +
                        age +
                        sex + 
                        weight + 
                        pool)

colnames(design) <- gsub("group", "", colnames(design))

contr.matrix = makeContrasts(
  CtlvI1 = Ctl - I1,
  CtlvI2 = Ctl - I2,
  CtlvI5 = Ctl - I5,
  I1vI2 = I1 - I2,
  I1vI5 = I1 - I5,
  I2vI5 = I2 - I5,
  levels = colnames(design))
```

### Mean-variance trend and sample weight plots
```{r}
v <- voomWithQualityWeights(dge, design, plot = TRUE)
```

### Fitting the model
```{r}
vfit <- lmFit(v, design) 
vfit <- contrasts.fit(vfit, contrasts = contr.matrix)
tfit <- treat(vfit, lfc = 1)
plotSA(tfit)
```

### Count DE genes
For a gene to be considered differentially expressed, we require a p-value of 0.1 with a false discovery rate correction and a log fold count difference of 1.
```{r}
dt <- decideTests(tfit, p.value = 0.1, adjust.method = "fdr")
summary(dt) %>% 
  kable() %>% 
  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"), full_width = F)

dt.tib <- as_tibble(dt, rownames = NA) %>% 
  rownames_to_column("gene") %>% 
  mutate_at(vars(starts_with("C")), as.numeric) %>% 
  mutate_at(vars(starts_with("I")), as.numeric) %>% 
  filter_at(vars(CtlvI1:I2vI5), any_vars(. != 0))
```

### Histogram of P-values
```{r}
hist(tfit$p.value, main = "P-values comparing groups (eBayes)")
```


### Volcano plot
```{r, warning = FALSE, message = FALSE, fig.height = 10, fig.width = 7, fig.align = "center"}
tmp1 <- topTreat(tfit, coef = 1, n = Inf)
results1 <- mutate(tmp1, sig=ifelse(tmp1$adj.P.Val<0.1, "Sig", "Not Sig"))
p1 <- ggplot(results1, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 1") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp2 <- topTreat(tfit, coef = 2, n = Inf)
results2 <- mutate(tmp2, sig=ifelse(tmp2$adj.P.Val<0.1, "Sig", "Not Sig"))
p2 <- ggplot(results2, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 2") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp3 <- topTreat(tfit, coef = 3, n = Inf)
results3 <- mutate(tmp3, sig=ifelse(tmp3$adj.P.Val<0.1, "Sig", "Not Sig"))
p3 <- ggplot(results3, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Control vs. Infected, Day 5") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp4 <- topTreat(tfit, coef = 4, n = Inf)
results4 <- mutate(tmp4, sig=ifelse(tmp4$adj.P.Val<0.1, "Sig", "Not Sig"))
p4 <- ggplot(results4, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Infected, Day 1 vs. Infected, Day 2") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp5 <- topTreat(tfit, coef = 5, n = Inf)
results5 <- mutate(tmp5, sig=ifelse(tmp5$adj.P.Val<0.1, "Sig", "Not Sig"))
p5 <- ggplot(results5, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Infected, Day 1 vs. Infected, Day 5") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")

tmp6 <- topTreat(tfit, coef = 6, n = Inf)
results6 <- mutate(tmp6, sig=ifelse(tmp6$adj.P.Val<0.1, "Sig", "Not Sig"))
p6 <- ggplot(results6, aes(logFC, -log10(P.Value))) +
  geom_point(aes(col = sig)) +
  scale_color_manual(values=c("black", "red")) +
  ggtitle("Infected, Day 2 vs. Infected, Day 5") +
  ylab("-Log10(Adjusted p-value)") +
  xlab("Log(Fold count)") +
  theme_bw() +
  theme(legend.position = "none")


grid.arrange(p1, p2, p3, p4, p5, p6, nrow = 3)
```


### Annotations for differentially expressed genes
#```{r message=FALSE, warning=FALSE}
#annot %>%
#  select(
#    gene_id,
#    sprot_Top_BLASTX_hit
#  ) %>%
#  rename(gene = gene_id) %>% 
#  inner_join(dt.tib) %>% 
#  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
#  separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>% 
#  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
#  mutate_at(.vars = vars(CtlvI1:I5vI14),
#            .funs = ~ ifelse(. == 1, "up", .)) %>% 
#  mutate_at(.vars = vars(CtlvI1:I5vI14),
#            .funs = ~ ifelse(. == -1, "down", .)) %>% 
#  distinct(gene, SwissProt_GeneName, .keep_all = TRUE) %>% 
#  select(-SwissProt_GeneFunction) %>% 
#  kable() %>%
#  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
#```

##### Gene functions
#```{r message=FALSE, warning=FALSE}
#annot %>%
#  select(
#    gene_id,
#    sprot_Top_BLASTX_hit,
#    Kegg
#  ) %>%
#  rename(gene = gene_id) %>% 
#  inner_join(dt.tib) %>% 
#  separate(sprot_Top_BLASTX_hit, into = c("sprot1", NA, NA, NA, NA, "sprot2", NA), "\\^") %>% 
#  separate(sprot2, sep = "=", into = c(NA, "SwissProt_GeneFunction")) %>% 
#  separate(sprot1, sep = "_", into = c("SwissProt_GeneName", NA)) %>% 
#  separate(Kegg, into = c("KEGG_ID", "KO_ID"), sep = "\\`") %>% 
#  distinct(gene, SwissProt_GeneName, KEGG_ID, .keep_all = TRUE) %>% 
#  select(gene, SwissProt_GeneName, SwissProt_GeneFunction, KEGG_ID, KO_ID) %>% 
#  filter(SwissProt_GeneName != ".") %>% 
#  kable() %>%
#  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
#```
#
#
#
#### Gene Ontology Enrichment Analysis
#```{r message=FALSE, warning=FALSE}
#topGO.mappings <- readMappings("G:/Shared drives/RNA Seq Supershedder Project/BWTE DE #manuscript/extData/GOdbGene.txt", sep = "\t", IDsep = ",")
#
#goEnrich <- function(targetComp, ...) {
#  DE.results <- dt.tib %>% filter(!!sym(targetComp) != 0)
#  
#  all.genes <- sort(unique(as.character(rownames(tfit$p.value))))
#  int.genes <- DE.results$gene
#  int.genes <- factor(as.integer(all.genes %in% int.genes))
#  names(int.genes) = all.genes
#  
#  go.obj.BP <- new("topGOdata", ontology='BP',
#              allGenes = int.genes,
#              annot = annFUN.gene2GO,
#              nodeSize = 5,
#              gene2GO = topGO.mappings)
#
#  go.obj.MF <- new("topGOdata", ontology='MF',
#                 allGenes = int.genes,
#                 annot = annFUN.gene2GO,
#                 nodeSize = 5,
#                 gene2GO = topGO.mappings)
#
#  go.obj.CC <- new("topGOdata", ontology='CC',
#                 allGenes = int.genes,
#                 annot = annFUN.gene2GO,
#                 nodeSize = 5,
#                 gene2GO = topGO.mappings)
#  
#  results.BP <- runTest(go.obj.BP, algorithm = "elim", statistic = "fisher")
#
#  results.tab.BP <- GenTable(object = go.obj.BP, 
#                          elimFisher = results.BP, 
#                          topNodes = 50) %>% 
#    as_tibble() %>% 
#    mutate(pVal = as.numeric(elimFisher)) %>% 
#    filter(pVal < 0.01) %>% 
#    mutate(Domain = "BP") %>% 
#    mutate(Comparison = targetComp)
#  
#  results.MF <- runTest(go.obj.MF, algorithm = "elim", statistic = "fisher")
#  
#  results.tab.MF <- GenTable(object = go.obj.MF, 
#                          elimFisher = results.MF, 
#                          topNodes = 50) %>% 
#    as_tibble() %>% 
#    mutate(pVal = as.numeric(elimFisher)) %>% 
#    filter(pVal < 0.01) %>% 
#    mutate(Domain = "MF") %>% 
#    mutate(Comparison = targetComp)
#  
#  results.CC <- runTest(go.obj.CC, algorithm = "elim", statistic = "fisher")
#  
#  results.tab.CC <- GenTable(object = go.obj.CC, 
#                          elimFisher = results.CC, 
#                          topNodes = 50) %>% 
#    as_tibble() %>% 
#    mutate(pVal = as.numeric(elimFisher)) %>% 
#    filter(pVal < 0.01) %>% 
#    mutate(Domain = "CC") %>% 
#    mutate(Comparison = targetComp)
#  
#  rbind(results.tab.BP,
#    results.tab.MF,
#    results.tab.CC)
#}
#
#results <- list()
#
#for(z in unique(names(dt.tib)[-1])) {
#  if (dt.tib %>% 
#      filter(!!sym(z) != 0) %>% 
#      nrow(.) > 0)
#  results[[length(results)+1]] <- goEnrich(z)
#}
#
#results.tib <- bind_rows(results, .id = "column_label") %>%
#  select(-column_label) %>% 
#  arrange(GO.ID, Comparison)
#
#kable(results.tib) %>%
#  kable_styling(bootstrap_options = c("striped", "hover", "condensed", "responsive"))
#```
